### PR TITLE
Code quality

### PR DESCRIPTION
**Pull request by rupertford**
All python is now pep8 compliant and is mostly pylint compliant. landscape reports 96%. Remaining pylint issues are:

src/inputs.py
R:113, 4: Too many arguments (6/5) (too-many-arguments)
I'm happy with the number of arguments in this piece of code.
C:159,12: Consider using enumerate instead of iterating with range and len (consider-using-enumerate)
I'm happy using range and len.

examples/PSyclone/psyclone.py
R:104, 4: Too many arguments (6/5) (too-many-arguments)
I'm happy with the number of arguments in this piece of code.
W:113,16: Access to a protected member _validate of a client class (protected-access)
This would require a change to a remote code, not one in this repository, so this error depends on external factors.

examples/shallow/execute.py
R: 40, 0: Too many local variables (18/15) (too-many-locals)
I'm happy with the number of variables in this piece of code (and the code will change in the future in any case)

landscape gives some errors that are not issues for pylint and pep8. For example it complains about the use of "_" and the variable not being used. I am not planning to change this as this is an accepted convention.

### CODE REVIEW

File: examples/PSyclone/psyclone.py [hunk 1]
```diff
@@ -60,29 +62,34 @@ def test_function(options, psy):
...
+    '''A specialisation of the melody Subsets input class providing all
+    possible combinations of module inlining for the particular code.'''
 
     def __init__(self, psy=None):
         self._psy = psy
         # get the kernels that can be inlined from PSyclone
         kernels = []
         kernel_names = []
-        for invoke in psy.invokes.invoke_list:
+        for invoke in self._psy.invokes.invoke_list:
```
Comment (arporter) [inline examples/PSyclone/psyclone.py:91]: `psy` could be None which would case this code to fail.

File: examples/PSyclone/psyclone.py [hunk 2]
```diff
@@ -100,32 +108,37 @@ def _recurse(self, siblings, my_index, options, all_options, invoke):
...
-        invokes = psy.invokes.invoke_list
+        invokes = my_psy.invokes.invoke_list
         #print "there are {0} invokes".format(len(invokes))
         if self._dependent_invokes:
-            print ("dependent invokes assumes fusion in one invoke might affect fusion "
-                   "in another invoke. This is not yet implemented")
+            print ("dependent invokes assumes fusion in one invoke might "
+                   "affect fusion in another invoke. This is not yet "
+                   "implemented")
             exit(1)
```
Comment (arporter) [inline examples/PSyclone/psyclone.py]: Should we raise an Exception here rather than just `exit()`?

File: examples/PSyclone/psyclone.py [hunk 3]
```diff
@@ -137,27 +150,29 @@ def options(self, psy):
...
-                        self._recurse(siblings, my_index, option, my_options, invoke)
+                        self._recurse(siblings, my_index, option, my_options,
+                                      invoke)
 
         return my_options
 
     @property
     def name(self):
-        ''' '''
+        '''Returns the name of the melody input class '''
```
Comment (arporter) [inline examples/PSyclone/psyclone.py]: Inconsistent white space in doc string (picky!)

File: examples/PSyclone/psyclone.py [hunk 4]
```diff
@@ -137,27 +150,29 @@ def options(self, psy):
...
 
-#FILE = "/home/rupert/proj/GungHo/PSyclone_trunk/examples/gocean/shallow_alg.f90"
-FILE = "/home/rupert/proj/GungHoSVN/PSyclone_trunk/examples/gocean/shallow_alg.f90"
-from parse import parse
-from psyGen import PSyFactory
-_, invoke_info = parse(FILE, api="gocean1.0")
-psy = PSyFactory("gocean1.0").create(invoke_info)
+#FILE = "/home/rupert/proj/GungHo/PSyclone_trunk/examples/gocean/
+#shallow_alg.f90"
+FILE = ("/home/rupert/proj/GungHoSVN/PSyclone_trunk/examples/"
```
Comment (arporter) [inline examples/PSyclone/psyclone.py]: Code uses a file that is not part of this project. I suggest copying that file into melody/examples.

File: examples/shallow/execute.py [hunk 5]
```diff
@@ -31,60 +31,72 @@
...
-    f.write(makefile_include)
-    f.flush()
+    my_file = open(location, 'w')
+    my_file.write(makefile_include)
+    my_file.flush()
 
     # compile shallow if required
     #import os
     #os.chdir("/home/rupert/proj/GOcean/shallow_benchmark/original")
+    base_path = "/home/rupert/proj/GOcean/shallow_benchmark/original"
```
Comment (arporter) [inline examples/shallow/execute.py]: Again, have a hard-wired path to GOcean code that is not a part of this project. I suggest adding a comment on where to get the GOcean code from and a variable to hold the path to its location on disk.

File: src/melody/inputs.py [hunk 6]
```diff
@@ -43,30 +46,35 @@ def __init__(self, name, options, state=False):
 
     @property
     def state(self):
-        ''' '''
+        '''Returns the state variable. This is False by default.'''
```
Comment (arporter) [inline src/melody/inputs.py]: How can a state be `False`? Should this be named something like `requires_state` instead or, if it is intended to hold a state object, maybe it should be None by default? (None is 'falsey' in python anyway so we keep that functionality if it is used anywhere.)

File: src/melody/inputs.py [hunk 7]
```diff
@@ -134,8 +154,8 @@ def __init__(self, name=None, inputs=None):
         Input.__init__(self, name, self._options)
 
     def _recurse(self, inputs, output, depth, max_depth):
-        ''' '''
-        if depth<max_depth:
+        '''We work out all combinations using this internal recursion class'''
```
Comment (arporter) [inline src/melody/inputs.py]: should that be "internal recusion method"?

**Review by arporter — CHANGES_REQUESTED** [review]
Generally good and I'm happy with what pep8 and pylint report. I've taken the opportunity to have a closer look at the lines of code involved in this merge request and have requested some changes - see the in-line comments/questions.

File: examples/PSyclone/psyclone.py (hunk 1 above)
Comment (rupertford) [reply examples/PSyclone/psyclone.py:91]: Quite right. I've made psy a required argument. Note this code is all subject to change as it is not working correctly at the moment so I've not bothered too much about being graceful with errors.

File: examples/PSyclone/psyclone.py (hunk 2 above)
Comment (rupertford) [reply examples/PSyclone/psyclone.py]: Again, this code does not currently work so I've not spent much time on its niceties. However, I've made it raise a RuntimeError exception.

File: examples/PSyclone/psyclone.py (hunk 3 above)
Comment (rupertford) [reply examples/PSyclone/psyclone.py]: Nothing wrong with looking for consistency. I've removed the extra white space.

File: examples/PSyclone/psyclone.py (hunk 4 above)
Comment (rupertford) [reply examples/PSyclone/psyclone.py]: OK, I've done this.

File: examples/shallow/execute.py (hunk 5 above)
Comment (rupertford) [reply examples/shallow/execute.py]: I've added the required benchmark code to the repository for the time being. I've tested it and it runs OK.

File: src/melody/inputs.py (hunk 6 above)
Comment (rupertford) [reply src/melody/inputs.py]: Agreed. I've set the default value to False and changed the documentation.

File: src/melody/inputs.py (hunk 7 above)
Comment (rupertford) [reply src/melody/inputs.py]: Yes. Changed.

**Comment by rupertford** [comment]
OK, suggested changes made. Ready for 2nd review. Note I also removed an additional warning that landscape brought up which was missed by my version of pylint.

**Review by arporter — APPROVED** [review]
Requested changes have been made. I didn't actually intend for you to import the GOcean code into melody but we can always change that later. Maybe we should move GOcean over to some git repo and then it could be done via submodules.

**Comment by jimboid** [comment]
This is the one I was just talking about Rupert.

**Pull request merged**